### ISSUE TITLE
Wait for node-critical pods before scheduling workload pods

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#7117]
**How to categorize this issue?**

/area robustness
/kind enhancement

## Motivation

When registering new `Nodes`, kubelet adds the `node.kubernetes.io/not-ready` taint to prevent scheduling workload pods to the `Node` until the `Ready` condition gets `True`.
However, the kubelet does not consider the readiness of node-critical pods.
Hence, the `Ready` condition might get `True` and the `node.kubernetes.io/not-ready` taint might get removed, for example, before the CNI daemon pod (e.g. `calico-node`) has successfully placed the CNI binaries on the machine.

We have seen cases, where a lot of workload pods get scheduled before `calico-node` got ready, and then a lot of images are pulled which exhausts the root disk’s IOPS, and then `calico-node` fails to get ready.
In such cases, the node is kind of broken, as pods can’t be created because of missing CNI binaries or other network problems, and the situation might only resolve an hour after node creation.

This problem has been discussed extensively in kubernetes, e.g., in https://github.com/kubernetes/kubernetes/issues/75890.
However, several proposals have been rejected because the problem can be solved by using the `--register-with-taints` kubelet flag and dedicated controllers ([ref](https://github.com/kubernetes/enhancements/pull/1003#issuecomment-619087019)).

Furthermore, the `CSINode` object might contain scheduling restrictions (e.g., the maximum number of volumes that can be attached to a given `Node`).
The `CSINode` object is only populated with these scheduling restrictions by kubelet once the CSI driver is successfully registered.
As long as the `CSINode` is not present or the driver has not been populated, kube-scheduler is not aware of potential scheduling restrictions and might perform scheduling in contradiction to these restrictions (ref https://github.com/kubernetes/kubernetes/issues/95911#issuecomment-717389872).

## Goal

- implement the solution, that was proposed in the linked kubernetes discussions, in gardener for all shoot clusters
- ensure that workload pods are only scheduled to `Nodes` once all node-critical pods got ready
- ensure that workload pods are only scheduled to `Nodes` once the corresponding `CSINode` object was created (if the `Node` is using CSI)
- make machine-controller-manager aware of the new checks and only continue rolling updates once all node-critical pods got ready in addition to waiting for the `Ready` condition (already done today)
- allow extensions and shoot owners to mark additional components as node-critical to instruct gardener to consider them

## Proposal

- gardenlet configures the shoot kubelets to register their `Node` object with a new taint:
	- add a new taint with key `node.gardener.cloud/critical-components-not-ready` and effect `NoSchedule` to `KubeletConfiguration.RegisterWithTaints`
	- by this, the `Node` object immediately contains this taint when registered by the kubelet
- a new well-known label is introduced: `node.gardener.cloud/critical-component`
    - all DaemonSets and Pods marked with this label are considered node-critical
    - only once all pods on the new node with this label get `Ready` the new taint is removed
- a new well-known annotation prefix is introduced: `node.gardener.cloud/wait-for-csi-node-`
    - `csi-driver-node` pods are marked with this annotation to inform gardener that a `Node` running such a pod is using the CSI driver name specified in the annotation value, e.g. `node.gardener.cloud/wait-for-csi-node-cinder=cinder.csi.openstack.org`
    - the annotation key suffix (`cinder` in the example above) can be chosen arbitrarily
    - a key prefix is used here to allow a single`csi-driver-node` pod to specify multiple CSI drivers that are registered by it (using the value for a fixed key would allow specifying only one CSI driver)
    - if gardener detects this annotation, it only removes the new taint from the `Node` object once a corresponding `CSINode` object has been created and the `spec.drivers` list has been populated with information about the driver with the name specified in the annotation value
- a new `node` controller is added to gardener-resource-manager:
    - the controller watches newly added `Node` objects and removes the taint once all node-critical pods are `Ready`
	- if the node still has the taint, it lists all `Pods` with the label and checks their readiness
	- in addition, it waits for all relevant `DaemonSets` to schedule a corresponding daemon pod
	    - for all `DaemonSets` that are labeled with `node.gardener.cloud/critical-component=true` and contain the label in `spec.template.metadata.labels`, it determines whether a corresponding daemon pod is already scheduled to the `Node` or if it must still be scheduled to the `Node`
	    - for this, it reuses the [scheduling helpers](https://github.com/kubernetes/kubernetes/blob/70617042976dc168208a41b8a10caa61f9748617/pkg/controller/daemon/daemon_controller.go#L1246-L1281) that are used by the `DaemonSet` controller for determining whether a given `Node` should run the daemon pod
	    - as long as there is no corresponding daemon pod, the controller does not remove the taint
	- in addition, it waits for the `CSINode` object to be created and populated with driver information
	    - if any of the node-critical pods are marked with the `wait-for-csi-node` annotation, the controller does not remove the taint as long as no `CSINode` object exists with the same name as the `Node` object or the `CSINode` object's `spec.drivers` list is missing the driver with the specified name
- gardenlet and extensions mark their node-critical DaemonSets and pods with the well-known label and annotation
	- examples of node-critical components managed by different gardener components:
		- gardenlet: kube-proxy, apiserver-proxy, node-local-dns
		- networking extension: CNI-related pods (e.g. `calico-node`)
		- provider extensions: csi-driver-node (also add the `node.gardener.cloud/wait-for-csi-node-<suffix>=<driver-name>` annotation)
	- all node-critical components need to tolerate `NoSchedule` taints (as before)
- gardenlet's shoot flows are adapted to ensure all critical system components are deployed before deploying the worker
	- this makes sure, the `node` controller knows about all node-critical pods and `DaemonSets` before checking whether the taint can be removed
- gardenlet passes all worker node labels to kubelet's `--node-label` flag
    - currently, machine-controller-manager is adding labels (specified by gardener or user) asynchronously after `Node` registration (ref https://github.com/gardener/machine-controller-manager/issues/740)
    - this would again lead to a race with the `node` controller's logic for determining whether daemon pods need to be scheduled to a given node (e.g., kube-proxy has a `nodeSelector` for the node label `worker.gardener.cloud/pool`)
    - passing all worker pool labels to `--node-labels` makes sure, the `Node` objects immediately contain the specified labels on registration
    - machine-controller-manager is still responsible for updating `Node` labels after registration according to specification changes

## Steps

- [x] https://github.com/gardener/gardener/pull/7426
- [x] https://github.com/gardener/gardener/pull/7406
    - API validation: deny adding the `node.gardener.cloud/critical-components-not-ready` taint to worker pools in the shoot spec
    - gardenlet: configure new taint in `KubeletConfiguration.RegisterWithTaints`
    - gardenlet: add new taint to `cluster-autoscaler` flag [`--ignore-taint`](https://github.com/kubernetes/autoscaler/blob/aa7733ce6803efe3be8a930e3901909c73d0ebbb/cluster-autoscaler/main.go#L183) (see https://github.com/gardener/gardener/issues/7117#issuecomment-1332624529)
    - gardener-resource-manager: introduce `node` controller to remove taint once all node-critical pods are scheduled and ready
    - integration test for `node` controller
    - e2e test for the new feature
    - add documentation for how extensions and users can make use of the feature
    - ensure all critical system components are deployed before deploying the worker
    - mark node-critical components managed by gardenlet with `node.gardener.cloud/critical-component`: kube-proxy, apiserver-proxy, node-local-dns
- [x] https://github.com/gardener/gardener/pull/7621
- [x] https://github.com/gardener/machine-controller-manager/pull/778
- mark node-critical components with `node.gardener.cloud/critical-component` (and optionally `node.gardener.cloud/wait-for-csi-node-<suffix>=<driver-name>`) in extensions
    - [x] [networking-calico](https://github.com/gardener/gardener-extension-networking-calico): https://github.com/gardener/gardener-extension-networking-calico/pull/236, released with [`v?`](https://github.com/gardener/gardener-extension-networking-calico/releases/tag/v?)
    - [x] [networking-cilium](https://github.com/gardener/gardener-extension-networking-cilium): https://github.com/gardener/gardener-extension-networking-cilium/pull/152, released with [`v?`](https://github.com/gardener/gardener-extension-networking-cilium/releases/tag/v?)
    - [x] [provider-alicloud](https://github.com/gardener/gardener-extension-provider-alicloud): https://github.com/gardener/gardener-extension-provider-alicloud/pull/567, https://github.com/gardener/gardener-extension-provider-alicloud/pull/582, released with [`v?`](https://github.com/gardener/gardener-extension-provider-alicloud/releases/tag/v?)
    - [x] [provider-aws](https://github.com/gardener/gardener-extension-provider-aws): https://github.com/gardener/gardener-extension-provider-aws/pull/701, https://github.com/gardener/gardener-extension-provider-aws/pull/716, released with [`v?`](https://github.com/gardener/gardener-extension-provider-aws/releases/tag/v?)
    - [x] [provider-azure](https://github.com/gardener/gardener-extension-provider-azure): https://github.com/gardener/gardener-extension-provider-azure/pull/632, https://github.com/gardener/gardener-extension-provider-azure/pull/643, released with [`v?`](https://github.com/gardener/gardener-extension-provider-azure/releases/tag/v?)
    - [x] [provider-gcp](https://github.com/gardener/gardener-extension-provider-gcp):  https://github.com/gardener/gardener-extension-provider-gcp/pull/555, https://github.com/gardener/gardener-extension-provider-gcp/pull/567, released with [`v?`](https://github.com/gardener/gardener-extension-provider-gcp/releases/tag/v?)
    - [x] [provider-metal](https://github.com/metal-stack/gardener-extension-provider-metal): https://github.com/metal-stack/gardener-extension-provider-metal/pull/285, https://github.com/metal-stack/duros-controller/pull/50, released with [`v?`](https://github.com/gardener/gardener-extension-provider-gcp/releases/tag/v?)
    - [x] [provider-openstack](https://github.com/gardener/gardener-extension-provider-openstack): https://github.com/gardener/gardener-extension-provider-openstack/pull/562, https://github.com/gardener/gardener-extension-provider-openstack/pull/583, released with [`v?`](https://github.com/gardener/gardener-extension-provider-openstack/releases/tag/v?)
    - [x] [provider-vsphere](https://github.com/gardener/gardener-extension-provider-vsphere): https://github.com/gardener/gardener-extension-provider-vsphere/pull/375, https://github.com/gardener/gardener-extension-provider-vsphere/pull/384, released with [`v0.24.0`](https://github.com/gardener/gardener-extension-provider-vsphere/releases/tag/v0.24.0)

**Comment by majst01** [comment]
This also affects the metal-stack.io components which are 

_EDIT(timebertt): added the metal-stack components to the list above_

**Comment by timebertt** [comment]
Thanks @majst01, I added the metal-stack components to the list :)

**Comment by vlerenc** [comment]
That means, these tolerations:

```
      tolerations:
      - effect: NoSchedule
        operator: Exists
      - effect: NoExecute
        operator: Exists
```

...for these components (at least `calico-typha`):

- `blackbox-exporter`
- `calico-kube-controllers`
- `calico-typha`

...cannot be removed, which we were discussing just a few days ago, because they cause trouble themselves/are wrong (these essential pods remain "sticky": even though the nodes break/are being drained).

W/o this PR and when we would have removed these tolerations, the node would turn `Ready` and these pods would only then be scheduled (not before), but with this PR, these tolerations must remain as for instance w/o `calico-typha` the `calico-node` pods won't turn `Ready` themselves.

That's somewhat of an issue now, no @istvanballok @DockToFuture @ScheererJ?

**Comment by timebertt** [comment]
I think there is some confusion involved here. Probably it's because I didn't fully understand what problems you have faced with the tolerations in system pods (ref https://github.com/gardener/gardener/issues/7103).

With this proposal, we won't change the `Ready` condition or the taints managed by kubelet (e.g., `node.kubernetes.io/not-ready`).
Hence, if the `NoExecute` toleration of `calico-typha` would be removed while still tolerating `NoSchedule` taints, it would still be scheduled to non-ready nodes, allowing `calico-node` to get ready.

Let's jump on a call in the next few days, so that I can understand what you are trying to achieve with https://github.com/gardener/gardener/issues/7103 / https://github.com/gardener/gardener/issues/7104 (cc @timuthy).
And please help me understand how this conflicts with this proposal so that I can adapt it accordingly.

**Comment by vlerenc** [comment]
Sorry, I am on sick leave, so not fully "present", but what I understood: You want to prevent nodes turning ready before the daemonset pods like calico-node turn ready, so that nothing gets scheduled to such half-way-through nodes, which won't happen if calico-typha is not running, which can only happen if either nodes are ready or it tolerates the related taints. Hence, catch22, no? But I agree, maybe we discuss that another time, when I can follow better.

As for what my comment was about: See above. Basically: let's run our system components/non-daemonset pods only on ready (or to-be-ready) not cordoned nodes, which is critical also for HA/recovery. These pods should not be stuck to nodes currently in distress (network partitioning or whatever) or cordoned/in drain (because of issues like mentioned above, or generally a rolling update or a targeted/purposeful node/zone evacuation). Today, these pods just keep stuck, because they tolerate way too much. So, if you can incorporate/combine the plan to remove those NoExecute/NoSchedule taints on those 3 system components (probably only the calico stuff is critical and the bbx never needed it), that would be great (tolerations for your planned new taints maybe).

**Comment by dguendisch** [comment]
I have one addition:
In https://github.com/kubernetes/kubernetes/issues/75890#issuecomment-1104723833 it's mentioned that the envisioned approach collides with cluster-autoscaler (as it would continue up-scaling until all critical pods are ready and the resp. well-known taint `node.gardener.cloud/critical-components-not-ready` got removed) and https://github.com/kubernetes/kubernetes/issues/75890#issuecomment-1224204404 describes a solution for this cluster-autoscaler issue. Hence I think part of this proposal needs to be to amend the cluster-autoscaler's `ignore-taint` flag with the proposed well-known taint `node.gardener.cloud/critical-components-not-ready`.

**Comment by timebertt** [comment]
Very good point. Thanks, @dguendisch! I added a corresponding task to the proposal 👍

**Comment by timebertt** [comment]
@timuthy and I checked in detail whether this proposal conflicts with the goal behind https://github.com/gardener/gardener/issues/7103 (also see https://github.com/gardener/gardener/issues/7103#issuecomment-1333855583).

In short, there is no conflict between this proposal and https://github.com/gardener/gardener/issues/7103.
And this proposal matches the expectations mentioned by @vlerenc in https://github.com/gardener/gardener/issues/7117#issuecomment-1332501363.

In detail, we currently try to realize the following general rules for system components:

- generally, system components should tolerate all `NoSchedule` taints (already done today)
	- should not tolerate all `NoExecute` taints to be part of taint-based eviction (differs from status-quo)
	- should tolerate all custom taints specified by the user in the shoot spec (accomplished by webhook as part of https://github.com/gardener/gardener/issues/7103)
- `calico-typha` is no exception to other system components, although it is needed for `calico-node` pods to get ready
	- might be scheduled even if  `node.gardener.cloud/critical-components-not-ready` taint is still present as it tolerates all `NoSchedule` taints (see above)
	- this means it will be scheduled even if node-critical pods are not ready yet
	- this is sufficient for `calico-node` pods to get ready for bootstrapping a new cluster (or waking up a hibernated one)
	- similar to other system components it should not tolerate all `NoExecute` taints to be part of taint-based eviction
- node-critical components should tolerate all `NoSchedule` and `NoExecute` taints (already done today)
	- should stay on the node even in the event of taint-based eviction


Thanks again for checking this with me, @timuthy! :)

**Comment by himanshu-kun** [comment]
I am currently reviewing the PR [778](https://github.com/gardener/machine-controller-manager/pull/778). I wanted to know do you also want to deal with situation where `calico-node` starts Crashing after the node is marked as viable candidate for scheduling workloads ? 
cc @unmarshall

**Comment by timebertt** [comment]
Right now, we only care about the bootstrapping process because of the reasons mentioned in the motivational section.
Taking care of crashing node-critical components during runtime is out of the scope of this proposal.

**Comment by timebertt** [comment]
Note: there has been a change to the proposal during the review: https://github.com/gardener/gardener/pull/7406#discussion_r1098630910
In addition to labeling node-critical pods with `node.gardener.cloud/critical-component=true`, the `DaemonSets` themselves also need to be labeled with `node.gardener.cloud/critical-component=true` for the `Node` controller to consider them.
cc @dergeberl

**Comment by SimonKienzler** [comment]
@majst01 @Gerrit91 as representatives of [provider-metal](https://github.com/metal-stack/gardener-extension-provider-metal): I'm currently adding a feature, #7621 , which allows the `csi-driver-node` pods to indicate that Gardener should not remove the `critical-components-not-ready` taint until the driver is registered in the `CSINode` object. Like with metal-stack/gardener-extension-provider-metal#285 , we also wanted to include your provider when adding the required annotation.

For all other providers, it was quite clear where to put the annotation (see e.g. gardener/gardener-extension-provider-openstack#583 ). For provider-metal, I'm not sure. Can you advise wether the `node.gardener.cloud/wait-for-csi-node` annotation makes sense for any of your node-critical pods (`speaker`, `csi-lvm-reviver` and `node-init`)?

**Comment by Gerrit91** [comment]
Hi @SimonKienzler,

thanks for asking! 😄

We have two different types of storage classes. One for csi-lvm (node local storage) and one for our NVMe over TCP volumes (Lightbits). A user can specify the default storage class through the shoot spec. 

- csi-lvm is a pretty old project and is not a "CSI driver" (we do not recommend it to our users but it's not possible to easily remove because people rely on it). It works by spawning pods on the nodes and it has no `CSINode` object, so the annotation does not fit there.
- The NVMe/TCP storage classes are deployed by a separate controller for which we only deploy a CRD through our extension-provider-metal. As the Lightbits CSI driver is way out of your domain, @mwennrich started a pull request here to add the annotation here: https://github.com/metal-stack/duros-controller/pull/50

**Comment by timebertt** [comment]
All steps of this issue have been completed. Thanks to everybody who contributed here! 🚀 
/close

**Comment by rfranzke** [comment]
Thank you, well done! :)
